Add explanation for Ubuntu Base and Minimal images (#218)
canonical/ubuntu-cloud-docs: all-clouds/all-clouds-explanation/ubuntu-base-and-minimal-images.rst

Ubuntu Base and Minimal Images
==============================

Broadly speaking, CPC produces two types of images - a base image and a minimal
image.

> 

Reviewer: Please provide links to where people can download these images or where they can discover them

Reviewer: Please also share that both Base and Minimal are available for all clouds too for all Ubuntu releases.

What are base and minimal images?
---------------------------------

Base images are those that are directly derived from Ubuntu Server-like images
and are meant for general consumption. On the other hand, minimal or minimized
images are those that are meant for machine-to-machine interactions, and have
some niceties meant for humans stripped out (such as manpages, translations,
editors, etc.). A great example would be our k8s images.

State of minimal images from Mantic onwards
-------------------------------------------

From Mantic Minotaur release onward (including Noble Numbat 24.04 LTS), we
worked through the minimal images to make it more efficient and truly minimal

> by constructing a “cloud-minimal” seed, which is nothing but the list of basic

Reviewer: please link to the cloud-minimal seed definition here.

packages that we need in a minimal cloud installation. The package list is then

> expanded by a program called “germinate”, which includes the dependencies of

Reviewer 2: ```suggestion
expanded by a program called `germinate`_, which includes the dependencies of
```

> the listed packages in the seed. Combined, we get the “image manifest”, aka the

Reviewer 2: ```suggestion
the listed packages in the seed. Combined, we get the `image manifest`_, i.e. the
```

> the listed packages in the seed. Combined, we get the “image manifest”, aka the

Reviewer: please link to the man pages for germinate here

Reviewer: please link to an example manifest and please refer to it as a package name and version manifest initially.

list of packages you’d expect to see in the minimal cloud image.

> 

Reviewer: An important detail on mantic+ is that the minimal image does not install any `Recommends` packages during image build.

> With that effort, we could reduce the package size from package count 426 to

Reviewer 2: ```suggestion
With this effort, we could reduce the package size from a package count of 426 to
```

> 288 (difference: 138) and also reduce the size of the minimal image. Taking

Reviewer: please share what these numbers are compared to - ie. the Jammy minimal image

download qcow2 images as an example, the image size reduced from 337.19MiB to
226.75MiB (difference: 110.44MiB).

> 

Reviewer: Please list the benefits of using minimal over base. Faster to boot, faster to deploy, faster to provision etc.

> What is unminimize? What does it do?

Reviewer 2: ```suggestion
What is ``unminimize``? What does it do?
```

------------------------------------

> Unminimize is essentially a script that’s shipped in /usr/bin by a package

Reviewer 2: ```suggestion
``unminimize`` is essentially a script that’s shipped in ``/usr/bin`` by a package
```

called “unminimize”. The goal of this unminimize script is to unminimize the

> minimal image and make it as close to that of a base image. It does this by

Reviewer 2: ```suggestion
minimal image and make it as close as possible to a base image. It does this by
```

re-enabling installation of all documentation in DPKG, restoring system
documentation and man pages, restoring system translations, and then finally

> installing some package on the top, like linux-image-virtual, et al. These

Reviewer 2: ```suggestion
installing some packages on the top, like linux-image-virtual, etc. These
```
The 'etc' instead of 'et al.' is just to reduce the number of spell check errors!

> packages make the human interaction easier as it installed the basic packages

Reviewer 2: ```suggestion
packages make human interaction easier as they include the installation of basic packages like editors.
```

> needed for so, like editors, et al.

Reviewer 2: delete this line.

To run ``unminimize``, you simply need to call:

> ``$ sudo unminimize``

Reviewer 2: Add the following (links) at the end of the file:

.. _`cloud-minimal`: https://ubuntu-archive-team.ubuntu.com/seeds/ubuntu.noble/cloud-minimal
.. _`germinate`: https://ubuntu-archive-team.ubuntu.com/germinate-output/ubuntu.noble/cloud-minimal
.. _`image manifest`: https://cloud-images.ubuntu.com/minimal/daily/noble/current/noble-minimal-cloudimg-amd64.manifest
